modules/stream-shredder/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/shredder/stream/Processing.scala

```diff
@@ -0,0 +1,143 @@
+      .flatMap(_.sink)  // Sinks must be issued sequentially
+      .compile
+      .drain
+  }
+
+  /**
+   * Get a callback that will be executed when window has been full written to destination
+   * The callback sends an SQS message and modifies the global state to reflect closed window
+   */
+  def getOnComplete[F[_]: Sync: Clock](compression: Compression, isTabular: SchemaKey => Boolean, root: URI, queueName: String)
+                                      (state: State.Windows[F])
+                                      (window: Window): F[Unit] = {
+    val find: State.WState => Boolean = {
+      case (w, status, _) => w == window && status == Status.Sealed
+    }
+    val update: State.WState => State.WState = {
+      case (w, _, state) => (w, Status.Closed, state)
+    }
+
+    state.modify(State.updateState(find, update, _._3)).flatMap { state =>
+      Completion.seal[F](compression, isTabular, root, queueName)(window, state)
+    }
+  }
+
+  /** Auto-incrementing sink ids (file suffix) */
+  def getSinkId[F[_]](windows: State.Windows[F])(window: Window): F[Int] =
+    windows.modify { stack =>
+      val update: State.WState => State.WState = {
+        case (w, s, state) => (w, s, state.copy(sinks = state.sinks + 1))
+      }
+      State.updateState(_._1 == window, update, _._3.sinks)(stack)
+    }
+
+  /** Build a `Stream` of parsed Snowplow events */
+  def getSource[F[_]: ConcurrentEffect: ContextShift](resources: Resources[F],
+                                                      config: Shredder.StreamInput): Stream[F, ParsedF[F]] =
+    config match {
+      case Shredder.StreamInput.Kinesis(appName, streamName, region, position) =>
+        sources.kinesis.read[F](appName, streamName, region, position)
+      case Shredder.StreamInput.File(dir) =>
+        sources.file.read[F](resources.blocker, dir)
+    }
+
+  /** Build a sink according to settings and pass it through [[generic.Partitioned]] */
+  def getSink[F[_]: ConcurrentEffect: ContextShift](blocker: Blocker,
+                                                    instanceId: String,
+                                                    config: Shredder.Output,
+                                                    sinkCount: Window => F[Int],
+                                                    onComplete: Window => F[Unit]): Grouping[F] =
+    config match {
+      case Shredder.Output(path, _) =>
+        val dataSink = Option(path.getScheme) match {
+          case Some("file") =>
+            file.getSink[F](blocker, path, config.compression, sinkCount) _
+          case Some("s3" | "s3a" | "s3n") =>
+            val (bucket, prefix) = S3.splitS3Path(S3.Folder.coerce(path.toString))
+            s3.getSink[F](bucket, prefix, config.compression, sinkCount, instanceId) _
+          case _ =>
+            val error = new IllegalArgumentException(s"Cannot create sink for $path. Possible options are file:// and s3://")
```

Comment:
If you change `getSink` to return `F[Grouping[F]]` then you could raise this error immediately in the `F` context, instead of returning a pipe that raises the error only when something gets piped through it.

Reply:
I thought about this quite a lot actually and decided to avoid type like `F[Stream[F, A]]` or `Resource[F, Stream[F, A]]` as `F[F[A]]`. Yes, sometimes they're more explicit, but my thinking was the user of the code will flatten them as a first step in 99% of cases.